clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -234,10 +234,15 @@ public RuntimeException lastError() {
     }
 
     public synchronized boolean ensurePartitionAdded(TopicPartition tp) {
-        if (isInTransaction() && !partitionsInTransaction.contains(tp)) {
-            transitionToFatalError(new IllegalStateException("Attempted to dequeue a record batch to send " +
-                    "for partition " + tp + ", which hasn't been added to the transaction yet"));
-            return false;
+        if (isInTransaction() || hasError()) {
```

Comment:
One quick question: what is the expected behavior when we are in a fatal error state? It seems we'd want to ensure that this function always returns false?